manifest-updater/manifest/storage.go

```diff
@@ -0,0 +1,260 @@
+	// WriteIngestorGlobalManifest writes the provided manifest to the writer's
+	// backing storage, or returns an error on failure.
+	WriteIngestorGlobalManifest(manifest IngestorGlobalManifest) error
+}
+
+// Fetcher fetches manifests from some storage
+type Fetcher interface {
+	// FetchDataShareProcessorSpecificManifest fetches the specific manifest for
+	// the specified data share processor and returns it, if it exists and is
+	// well-formed. Returns (nil,  nil) if the  manifest does not exist.
+	// Returns (nil, error) if something went wrong while trying to fetch or
+	// parse the manifest.
+	FetchDataShareProcessorSpecificManifest(dataShareProcessorName string) (*DataShareProcessorSpecificManifest, error)
+	// IngestorGlobalManifestExists returns true if the global manifest exists
+	// and is well-formed. Returns (false, nil) if it does not exist. Returns
+	// (false, error) if something went wrong while trying to fetch or parse the
+	// manifest.
+	IngestorGlobalManifestExists() (bool, error)
+}
+
+// Bucket specifies the cloud storage bucket where manifests are stored
+type Bucket struct {
+	// Bucket is the name of the bucket, without any URL scheme
+	Bucket string `json:"bucket"`
+	// AWSRegion is the region the bucket is in, if it is an S3 bucket
+	AWSRegion string `json:"aws_region,omitempty"`
+	// AWSProfile is the AWS CLI config profile that should be used to
+	// authenticate to AWS, if the bucket is an S3 bucket
+	AWSProfile string `json:"aws_profile,omitempty"`
+}
+
+// NewStorage creates an instance of the appropriate implementation of Writer for
+// the provided bucket
+func NewStorage(bucket *Bucket) (Storage, error) {
+	if bucket.AWSRegion != "" {
+		return newS3(bucket.Bucket, bucket.AWSRegion, bucket.AWSProfile)
+	}
+	return newGCS(bucket.Bucket)
+}
+
+// GCSStorage is a Storage that stores manifests in a Google Cloud Storage bucket
+type GCSStorage struct {
+	client                 *storage.Client
+	manifestBucketLocation string
+}
+
+// newGCS creates a GCSStorage
+func newGCS(manifestBucketLocation string) (*GCSStorage, error) {
+	client, err := storage.NewClient(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("unable to create a new storage client from background credentials: %w", err)
+	}
+	return &GCSStorage{client, manifestBucketLocation}, nil
+}
+
+func (s *GCSStorage) WriteIngestorGlobalManifest(manifest IngestorGlobalManifest) error {
+	return s.writeManifest(manifest, "global-manifest.json")
+}
+
+func (s *GCSStorage) WriteDataShareProcessorSpecificManifest(manifest DataShareProcessorSpecificManifest, dataShareProcessorName string) error {
+	return s.writeManifest(manifest, fmt.Sprintf("%s-manifest.json"))
+}
+
+func (s *GCSStorage) writeManifest(manifest interface{}, path string) error {
+	log.Info().
+		Str("path", path).
+		Msg("writing a manifest file")
+
+	ioWriter := s.getWriter(path)
+
+	if err := json.NewEncoder(ioWriter).Encode(manifest); err != nil {
+		_ = ioWriter.Close()
```

Comment:
consider not throwing away this error and instead always appending it beyond the first `err` even if it's `nil`. I don't know about this particular writer, but some do give different and yet still useful errors on Close for network failures that caused the earlier Write to fail, particularly if over UDP. But again, feel free to ignore if I'm offbase here.

Reply:
No, I think you raise a good point. We have to be more careful than usual with `cloud.google.com/go/storage.Writer` because it doesn't actually write to the network until `Writer.Close()` is called, so if you ignore the error from `Close()` you might not learn about failed writes. In this case we've already failed but it's still good to be meticulous about error handling all the time.